### PR TITLE
Functions are cleaner than subclasses

### PR DESCRIPTION
**Pull request by DavidVujic**
A drawback with a class extending the Array global, is that you are required to instantiate it using the 'new' keyword. Also, the 'diff' feature is limited to the subclass only.

### CODE REVIEW

**Comment by IAmAdamTaylor** [comment]
Agree with this due to the ability to create array literals. Arrays created by JSON.parse() also wouldn't get the new functionality. 

Should the function name be diffArray() to express that specifically works with arrays?

File: README.md [hunk 1]
```diff
@@ -609,6 +609,35 @@ class SuperArray extends Array {
 ```
 **[⬆ back to top](#table-of-contents)**
 
+### Functions are often cleaner than subclasses ###
```
Comment (ryanmcdermott) [inline README.md]: Does this replace the `SuperArray` example?

File: README.md [hunk 2]
```diff
@@ -609,6 +609,35 @@ class SuperArray extends Array {
...
+Classes are good when you need state. Consider writing simple functions if state is not an issue. A drawback with a class extending the Array global, is that you are required to instantiate it using the 'new' keyword. Normally an array is created with:
+
+```javascript
+const arr = []; 
+```
+
+A class will limit the usage of the diff feature to the subclass only. A better option is to write a stateless function, that accept two arrays and calculate the difference between them.
+
+```javascript
```
Comment (ryanmcdermott) [inline README.md]: This doesn't keep with the format of Bad/Good.

**Comment by DavidVujic** [comment]
Thank you for the feedback @ryanmcdermott! I will update the pull request!

**Comment by DavidVujic** [comment]
In this PR, I have added a section and did not remove the SuperArray example. Maybe the SuperArray section should be rewritten instead? I think the idea of avoiding adding code to globals is clean code. Maybe a different pull request?

File: README.md (hunk 2 above)
Comment (DavidVujic) [inline README.md]: Fixed!

File: README.md (hunk 1 above)
Comment (DavidVujic) [inline README.md]: I would instead recommend a rewrite of the SuperArray example. Will do another PR with rewitten section.

**Comment by vsemozhetbyt** [comment]
@DavidVujic Sorry, the original code was changed in https://github.com/ryanmcdermott/clean-code-javascript/pull/124.

**Comment by DavidVujic** [comment]
Great @vsemozhetbyt! I saw the changes, very nice improvement.

**Comment by vsemozhetbyt** [comment]
Thank you) I mean, maybe it would be good for consistency to update these fragments in your examples as well.

**Comment by DavidVujic** [comment]
Done!

Now hoping for a merge and/or more feedback for improvement.

**Comment by ryanmcdermott** [comment]
I think this is a really great idea for an addition! The only issue I have is that `SuperArray` holds state in that the primary state is the array data itself. Is there another simple example we can use? I don't want to contradict with the other use of `SuperArray`. I've had some consistency problems in this project that are almost fully resolved 😄

**Comment by DavidVujic** [comment]
Thank you for the feedback @ryanmcdermott! I have given it some thought and think the "good" part of the SuperArray example should be a stateless function, because of some drawbacks that I think are major:
* complex syntax: `const arr = new SuperArray();` instead of the much cleaner shorthand way: `const arr = [];`
* diff is a feature that shouldn't require any state.
* only instances of SuperArray (or extensions of that class) can use the Array diff function.

Let me know what you think! If you don't like it, please tell me and I will close the pull request.

**Comment by ryanmcdermott** [comment]
Hey David! Thanks for spending time revising this. I want to combine this PR with another: https://github.com/ryanmcdermott/clean-code-javascript/pull/139

The motivation is that I think we should say something about the nature of classes in the preamble of the _Classes_ section. We can incorporate your thoughts into that introduction. I want to leave this example the way it is for now! If you have time would love to see your work as an intro in the _Classes_ section!
